woocommerce/woocommerce
# Add an API endpoint to disconnect from WooCommerce.com

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Closes # .



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Make sure the site is connected to WooCommerce.com.
2. For the test instructions it is assumed that your WC test site URL is `wc-core.test:8888`. Please replace it every where with the actual URL of the site used for testing.
3. Find the site access token by running `wp option get woocommerce_helper_data` on CLI for your test site.
```
array (
  'auth' =>
  array (
    'access_token' => 'd617b3852d114827a1fbd245c67...',
    'access_token_secret' => '4edd37b2cb0fcea8e831bfaf...',
    'site_id' => 9,
    'user_id' => 1,
    'updated' => 1733412964,
  ),
)
```
4. To get a request signature run the following in a php shell. (replace the `host` if needed, no need to change `hash`)
```
$payload = array(
	'host' => 'wc-core.test:8888',
	'request_uri' => '/?rest_route=/wccom-site/v3/connection/disconnect',
	'method' => 'POST',
	'body' => '{"hash":"3ac105e16f23978b7d3cad9070fbc5adab7375ccebc0d9f2f959ef709319f6c8"}'
);

echo hash_hmac( 'sha256', json_encode($payload), '<`access_token_secret` from step 3>');
```
5. Make bellow request:
```
curl -v -X POST 'http://wc-core.test:8888/?rest_route=/wccom-site/v3/connection/disconnect' \
--header 'X-Woo-Signature: <signature from step 4>' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer <access token from step 3>' \
--data-raw '{"hash":"3ac105e16f23978b7d3cad9070fbc5adab7375ccebc0d9f2f959ef709319f6c8"}'
```
6. You should get a 403 error and site will not be disconnected.
7. Disable the request hash verification by returning `true` at the top of function [\WC_Helper::verify_request_hash](https://github.com/woocommerce/woocommerce/blob/22563c3c26d1fc499df4c55e17a3fa4e80e71555/plugins/woocommerce/includes/admin/helper/class-wc-helper.php#L1617)
8. Run the command in step 6 again.
9. The site should be disconnected.


### Changelog entry












<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

Add an API endpoint to disconnect a site from WooCommerce.com

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
